File: tests/AcceptanceTests/StepDefinitions/NavigationSteps.cs
// ============================================================================
// RAJ Financial - Navigation Step Definitions
// ============================================================================
// Reqnroll step definitions for Navigation.feature
//
// Role Model:
//   - Client: Standard user who owns their financial data
//   - Administrator: Platform staff with system-wide access
//
// Fine-grained access control (who can see whose data) is handled via
// DataAccessGrant entities at the API level, not through additional roles.
// ============================================================================

using Microsoft.Playwright;
using RajFinancial.AcceptanceTests.Hooks;
using Reqnroll;

namespace RajFinancial.AcceptanceTests.StepDefinitions;

/// <summary>
///     Step definitions for navigation scenarios.
/// </summary>
[Binding]
public class NavigationSteps(ScenarioContext scenarioContext)
{
    /// <summary>
    ///     Test user emails by role (not sensitive - can be hardcoded).
    ///     Note: Only Client and Administrator roles are defined.
    ///     Data sharing is handled via DataAccessGrant entities.
    /// </summary>
    private static readonly Dictionary<string, string> testUserEmails = new(StringComparer.OrdinalIgnoreCase)
    {
        ["Client"] = "test-client@rajfinancialdev.onmicrosoft.com",
        ["Administrator"] = "test-admin@rajfinancialdev.onmicrosoft.com"
    };

    private IPage Page => scenarioContext.GetPage();

    [Given(@"I am logged in as a ""(.*)""")]
    [Given(@"I am logged in as an ""(.*)""")]
    public async Task GivenIAmLoggedInAs(string role)
    {
        scenarioContext.Set(role, "UserRole");

        var email = GetEmailForRole(role);
        scenarioContext.Set(email, "LoggedInEmail");

        if (await TryLoginWithStorageStateAsync(role, email)) return;

        var password = GetPasswordForRole(role);

        if (string.IsNullOrEmpty(password))
            throw new InconclusiveException(
                $"Test password not configured for role '{role}'. " +
                $"Set password in appsettings.local.json or TEST_{role.ToUpper()}_PASSWORD environment variable.");

        await PerformInteractiveLoginAsync(email, password);
    }

    private static string GetEmailForRole(string role)
    {
        if (!testUserEmails.TryGetValue(role, out var email))
            throw new ArgumentException(
                $"Unknown test role: '{role}'. Valid roles: {string.Join(", ", testUserEmails.Keys)}");
        return email;
    }

    private static string? GetPasswordForRole(string role)
    {
        return TestConfiguration.Instance.GetPassword(role)
               ?? Environment.GetEnvironmentVariable($"TEST_{role.ToUpper()}_PASSWORD");
    }

    private async Task<bool> TryLoginWithStorageStateAsync(string role, string email)
    {
        var storageStatePath = TestConfiguration.Instance.GetStorageStatePath(role);
        if (string.IsNullOrWhiteSpace(storageStatePath)) return false;

        var storageValid = await PlaywrightHooks.ValidateOrRegenerateStorageStateAsync(role, email, storageStatePath);
        if (!storageValid) return false;

        var currentViewport = await CloseExistingPageAndGetViewportAsync();
        await CloseExistingContextAsync();

        var context = await PlaywrightHooks.Browser.NewContextAsync(new BrowserNewContextOptions
        {
            StorageStatePath = storageStatePath,
            ViewportSize = currentViewport ?? new ViewportSize { Width = 1280, Height = 720 }
        });
        var page = await context.NewPageAsync();

        scenarioContext.Set(context, "BrowserContext");
        scenarioContext.Set(page, "Page");

        await page.GotoAsync(PlaywrightHooks.BaseUrl, new PageGotoOptions { WaitUntil = WaitUntilState.NetworkIdle });
        await page.WaitForTimeoutAsync(1000);

        await WaitForAuthenticatedState(page);
        return true;
    }

    private async Task<ViewportSize?> CloseExistingPageAndGetViewportAsync()
    {
        if (!scenarioContext.TryGetValue<IPage>("Page", out var existingPage)) return null;

        ViewportSize? currentViewport = null;
        var pageViewport = existingPage.ViewportSize;
        if (pageViewport != null)
            currentViewport = new ViewportSize { Width = pageViewport.Width, Height = pageViewport.Height };
        await existingPage.CloseAsync();
        return currentViewport;
    }

    private async Task CloseExistingContextAsync()
    {
        if (scenarioContext.TryGetValue<IBrowserContext>("BrowserContext", out var existingContext))
            await existingContext.CloseAsync();
    }

    private async Task PerformInteractiveLoginAsync(string email, string password)
    {
        await Page.GotoAsync(PlaywrightHooks.BaseUrl);
        await Page.WaitForTimeoutAsync(2000);

        // Look for Sign In button (updated from old "Sign In / Sign Up" text)
        var loginButton = Page.Locator("text=Sign In").First;
        if (!await loginButton.IsVisibleAsync())
            // Fallback to Get Started button or href-based selector
            loginButton = Page.Locator("a[href*='authentication/login'], a[href*='authentication/register']").First;

        if (await loginButton.IsVisibleAsync())
        {
            await loginButton.ClickAsync();

            await Page.WaitForURLAsync(url =>
                    url.Contains("ciamlogin.com") ||
                    url.Contains("login.microsoftonline.com") ||
                    url.Contains("b2clogin.com"),
                new PageWaitForURLOptions { Timeout = 15000 });

            await PlaywrightHooks.HandleEntraLoginPage(Page, email, password);

            await WaitForAuthenticatedState(Page);
        }
    }

    /// <summary>
    ///     Waits for the authenticated state to be resolved and any redirects to complete.
    /// </summary>
    private static async Task WaitForAuthenticatedState(IPage page)
    {
        // Wait for Blazor to process the auth callback and update state
        await page.WaitForLoadStateAsync(LoadState.NetworkIdle);

        // Wait for either:
        // 1. Redirect to a dashboard (admin, advisor, or client)
        // 2. Log out button to appear (indicating auth is complete)
        // 3. Navigation sidebar to appear (authenticated layout)
        var maxAttempts = 30; // 15 seconds total
        for (var i = 0; i < maxAttempts; i++)
        {
            await page.WaitForTimeoutAsync(500);

            var url = page.Url;
            if (url.Contains("/dashboard") || url.Contains("/admin") || url.Contains("/advisor") ||
                url.Contains("/client")) break;

            // Check for authenticated UI elements
            var logoutVisible = await page.Locator("text=/Log out/i").First.IsVisibleAsync();
            var sidebarVisible =
                await page.Locator(".raj-sidebar, nav[aria-label*='navigation']").First.IsVisibleAsync();

            if (logoutVisible || sidebarVisible) break;
        }

        // Final wait for page to fully stabilize
        await page.WaitForLoadStateAsync(LoadState.NetworkIdle);
        await page.WaitForTimeoutAsync(1000);
    }

    [Given(@"I am not logged in")]
    public async Task GivenIAmNotLoggedIn()
    {
        // Navigate to the app without logging in
        // Clear any existing auth state by navigating to logout first
        await Page.GotoAsync(PlaywrightHooks.BaseUrl + "/authentication/logout");
        await Page.WaitForTimeoutAsync(1000);

        // Navigate to home
        await Page.GotoAsync(PlaywrightHooks.BaseUrl);
        await Page.WaitForTimeoutAsync(1000);
    }

    [When(@"I navigate to ""(.*)""")]
    public async Task WhenINavigateTo(string path)
    {
        var fullUrl = PlaywrightHooks.BaseUrl.TrimEnd('/') + path;
        await Page.GotoAsync(fullUrl, new PageGotoOptions { WaitUntil = WaitUntilState.NetworkIdle });

        // Wait for Blazor to fully render
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);

        // Additional wait for Blazor components to initialize
        await Page.WaitForTimeoutAsync(2000);

        // Wait for main content to be visible
        try
        {
            await Page.WaitForSelectorAsync("main, [role='main'], .raj-content",
                new PageWaitForSelectorOptions { Timeout = 5000, State = WaitForSelectorState.Visible });
        }
        catch
        {
            // Main content selector not found, continue anyway
        }
    }

    [Then(@"I should not see an access denied message")]
    public async Task ThenIShouldNotSeeAnAccessDeniedMessage()
    {
        var content = await Page.ContentAsync();
        Assert.DoesNotContain("Access Denied", content);
        Assert.DoesNotContain("not authorized", content.ToLower());
    }

    [Then(@"I should see an access denied message")]
    public async Task ThenIShouldSeeAnAccessDeniedMessage()
    {
        var content = await Page.ContentAsync();
        Assert.True(
            content.Contains("Access Denied") || content.Contains("not authorized", StringComparison.OrdinalIgnoreCase),
            "Should see an access denied message");
    }

    [When(@"I view the navigation menu")]
    public async Task WhenIViewTheNavigationMenu()
    {
        // Wait for the page to be fully loaded after login
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
        await Page.WaitForTimeoutAsync(1000);

        // Wait for sidebar/navigation to be present
        try
        {
            await Page.WaitForSelectorAsync(".raj-sidebar, .nav-menu, nav",
                new PageWaitForSelectorOptions { Timeout = 5000, State = WaitForSelectorState.Visible });
        }
        catch
        {
            // Navigation might not be visible yet, take a screenshot for debugging
            var screenshotPath = Path.Combine(Path.GetTempPath(), $"nav-debug-{DateTime.Now:yyyyMMdd-HHmmss}.png");
            await Page.ScreenshotAsync(new PageScreenshotOptions { Path = screenshotPath, FullPage = true });
            Console.WriteLine($"Debug screenshot saved to: {screenshotPath}");
        }

        // On mobile, we might need to open the hamburger menu first
        var viewport = Page.ViewportSize;
        if (viewport != null && viewport.Width < 768)
        {
            var hamburger = Page.Locator(".raj-menu-toggle, [aria-label*='menu'], .hamburger, .menu-toggle").First;
            if (await hamburger.IsVisibleAsync())
            {
                await hamburger.ClickAsync();
                await Page.WaitForTimeoutAsync(500);
            }
        }
    }

    [Then(@"I should see the ""(.*)"" link in admin section")]
    public async Task ThenIShouldSeeTheLinkInAdminSection(string linkText)
    {
        // Look for the link within the admin navigation section
        var adminSection = Page.Locator("nav").Filter(new LocatorFilterOptions { HasText = "Administration" });
        var link = adminSection.Locator($"text={linkText}");
        await Assertions.Expect(link).ToBeVisibleAsync();
    }

    [When(@"I press Tab multiple times")]
    public async Task WhenIPressTabMultipleTimes()
    {
        for (var i = 0; i < 5; i++)
        {
            await Page.Keyboard.PressAsync("Tab");
            await Page.WaitForTimeoutAsync(100);
        }
    }

    [Then(@"I should be able to navigate through menu items")]
    public async Task ThenIShouldBeAbleToNavigateThroughMenuItems()
    {
        var activeElement = await Page.EvaluateAsync<string>(
            "document.activeElement?.tagName");
        Assert.NotNull(activeElement);
        Assert.NotEqual("BODY", activeElement.ToUpper());
    }

    [Then(@"I should see a hamburger menu button")]
    public async Task ThenIShouldSeeAHamburgerMenuButton()
    {
        // Hamburger menu only visible on mobile for authenticated users
        var hamburger = Page.Locator(".raj-menu-toggle").First;
        await Assertions.Expect(hamburger).ToBeVisibleAsync();
    }

    [When(@"I click the hamburger menu button")]
    public async Task WhenIClickTheHamburgerMenuButton()
    {
        var hamburger = Page.Locator(".raj-menu-toggle").First;
        await hamburger.ClickAsync();
        // Wait for sidebar animation to complete
        await Page.WaitForTimeoutAsync(300);
    }

    [Then(@"the navigation menu should be visible")]
    public async Task ThenTheNavigationMenuShouldBeVisible()
    {
        // Check if the sidebar has the "open" class (mobile) or is visible (desktop)
        var sidebar = Page.Locator(".raj-sidebar");
        await Assertions.Expect(sidebar.First).ToBeVisibleAsync();
    }
}

File: tests/e2e/step-definitions/navigation.steps.ts
// ============================================================================
// Navigation Step Definitions
// ============================================================================
// Steps for Navigation.feature — tests role-based sidebar navigation.
// ============================================================================

import { Given, When, Then } from "@cucumber/cucumber";
import { expect } from "@playwright/test";
import { CustomWorld } from "../support/world";
import { config } from "../support/config";
import {
  handleEntraLogin,
  waitForEntraLoginPage,
} from "../support/helpers/entra-auth";
import { getBrowser, getStorageStatePath } from "../support/hooks";

const testUserEmails: Record<string, string> = {
  Client: "test-client@rajfinancialdev.onmicrosoft.com",
  Administrator: "test-admin@rajfinancialdev.onmicrosoft.com",
};

// ── Login steps ─────────────────────────────────────────────────────────────

Given(
  "I am logged in as a {string}",
  { timeout: 60_000 },
  async function (this: CustomWorld, role: string) {
    await loginAsRole.call(this, role);
  }
);

Given(
  "I am logged in as an {string}",
  { timeout: 60_000 },
  async function (this: CustomWorld, role: string) {
    await loginAsRole.call(this, role);
  }
);

async function loginAsRole(this: CustomWorld, role: string) {
  this.set("UserRole", role);

  const email = testUserEmails[role];
  if (!email) {
    throw new Error(
      `Unknown role: '${role}'. Valid: ${Object.keys(testUserEmails).join(", ")}`
    );
  }
  this.set("LoggedInEmail", email);

  // Try cached storage state first (pre-generated in BeforeAll).
  // This avoids hitting Entra on every scenario — mirrors C# PlaywrightHooks.
  const cachedState = getStorageStatePath(role);
  if (cachedState) {
    const fs = await import("fs");
    if (fs.existsSync(cachedState)) {
      const viewport = this.page.viewportSize() ?? {
        width: 1280,
        height: 720,
      };
      await this.page.close();
      await this.context.close();

      this.context = await getBrowser().newContext({
        storageState: cachedState,
        viewport,
        ignoreHTTPSErrors: true,
      });
      this.page = await this.context.newPage();

      await this.page.goto(config.baseUrl, { waitUntil: "networkidle" });
      await this.page.waitForTimeout(1000);

      // Verify we're actually authenticated
      const logoutVisible = await this.page
        .locator("text=/Log out/i")
        .first()
        .isVisible()
        .catch(() => false);
      const onDashboard = this.page.url().includes("/dashboard");

      if (logoutVisible || onDashboard) return;
      // Storage state stale — fall through to interactive login
      console.warn(`⚠️ Cached state for ${role} is stale — falling back to interactive login`);
    }
  }

  // Also try user-configured storage state path from .env
  const storageStatePath = config.testUsers[role]?.storageStatePath;
  if (storageStatePath) {
    const fs = await import("fs");
    if (fs.existsSync(storageStatePath)) {
      const viewport = this.page.viewportSize() ?? {
        width: 1280,
        height: 720,
      };
      await this.page.close();
      await this.context.close();

      this.context = await getBrowser().newContext({
        storageState: storageStatePath,
        viewport,
        ignoreHTTPSErrors: true,
      });
      this.page = await this.context.newPage();

      await this.page.goto(config.baseUrl, { waitUntil: "networkidle" });
      await this.page.waitForTimeout(1000);

      const logoutVisible = await this.page
        .locator("text=/Log out/i")
        .first()
        .isVisible()
        .catch(() => false);
      const onDashboard = this.page.url().includes("/dashboard");

      if (logoutVisible || onDashboard) return;
    }
  }

  // Interactive login — last resort
  const password =
    config.testUsers[role]?.password ??
    process.env[`TEST_${role.toUpperCase()}_PASSWORD`];

  if (!password) {
    throw new Error(
      `Password not configured for role '${role}'. ` +
        `Set TEST_${role.toUpperCase()}_PASSWORD in .env or environment.`
    );
  }

  await this.page.goto(config.baseUrl);
  await this.page.waitForTimeout(1000);

  // Click Sign In — try multiple selectors since mobile/desktop render
  // different buttons. Use CSS visibility to find the one actually shown.
  const loginSelectors = [
    "button:visible:has-text('Sign In')",
    "a:visible:has-text('Sign In')",
    "button:has-text('Sign In')",
    "a[href*='authentication/login']",
    "a[href*='authentication/register']",
  ];

  let clicked = false;
  for (const selector of loginSelectors) {
    try {
      const btn = this.page.locator(selector).first();
      if (await btn.isVisible()) {
        await btn.click();
        clicked = true;
        break;
      }
    } catch {
      // try next
    }
  }

  if (!clicked) {
    throw new Error(
      `Sign In button not found on ${this.page.url()}. ` +
        `Viewport: ${JSON.stringify(this.page.viewportSize())}`
    );
  }

  await waitForEntraLoginPage(this.page);
  await handleEntraLogin(this.page, email, password);
  await waitForAuthenticatedState(this.page);
}

Given("I am not logged in", async function (this: CustomWorld) {
  // Clear any existing auth state
  await this.page.goto(config.baseUrl);
  await this.page.evaluate(() => {
    localStorage.clear();
    sessionStorage.clear();
  });
  await this.page.reload({ waitUntil: "networkidle" });
  await this.page.waitForTimeout(1000);
});

// ── Navigation steps ────────────────────────────────────────────────────────

When(
  "I view the navigation menu",
  async function (this: CustomWorld) {
    await this.page.waitForLoadState("networkidle");
    await this.page.waitForTimeout(1000);

    // On mobile, open hamburger menu
    const viewport = this.page.viewportSize();
    if (viewport && viewport.width < 768) {
      const hamburger = this.page
        .locator(
          "[data-testid='mobile-menu-toggle'], [aria-label*='menu' i]"
        )
        .first();
      if (await hamburger.isVisible()) {
        await hamburger.click();
        await this.page.waitForTimeout(500);
      }
    }
  }
);

Then(
  "I should see the {string} link in admin section",
  async function (this: CustomWorld, linkText: string) {
    const adminSection = this.page
      .locator("nav")
      .filter({ hasText: "Administration" });
    const link = adminSection.locator(`text=${linkText}`);
    await expect(link).toBeVisible();
  }
);

When(
  "I press Tab multiple times",
  async function (this: CustomWorld) {
    for (let i = 0; i < 5; i++) {
      await this.page.keyboard.press("Tab");
      await this.page.waitForTimeout(100);
    }
  }
);

Then(
  "I should be able to navigate through menu items",
  async function (this: CustomWorld) {
    const tag = await this.page.evaluate(
      () => document.activeElement?.tagName
    );
    expect(tag).not.toBeNull();
    expect(tag!.toUpperCase()).not.toBe("BODY");
  }
);

Then(
  "I should see a hamburger menu button",
  async function (this: CustomWorld) {
    const hamburger = this.page
      .locator("[data-testid='mobile-menu-toggle']")
      .first();
    await expect(hamburger).toBeVisible();
  }
);

When(
  "I click the hamburger menu button",
  async function (this: CustomWorld) {
    const hamburger = this.page
      .locator("[data-testid='mobile-menu-toggle']")
      .first();
    await hamburger.click();
    await this.page.waitForTimeout(300);
  }
);

Then(
  "the navigation menu should be visible",
  async function (this: CustomWorld) {
    const sidebar = this.page.locator("[data-testid='sidebar']");
    await expect(sidebar.first()).toBeVisible();
  }
);

// ── Helpers ─────────────────────────────────────────────────────────────────

async function waitForAuthenticatedState(page: import("playwright").Page) {
  await page.waitForLoadState("networkidle");

  const maxAttempts = 30;
  for (let i = 0; i < maxAttempts; i++) {
    await page.waitForTimeout(500);

    const url = page.url();
    if (
      url.includes("/dashboard") ||
      url.includes("/admin") ||
      url.includes("/advisor")
    ) {
      break;
    }

    const logoutVisible = await page
      .getByText(/log out/i)
      .first()
      .isVisible()
      .catch(() => false);
    const sidebarVisible = await page
      .locator("[data-testid='sidebar'], nav[aria-label]")
      .first()
      .isVisible()
      .catch(() => false);

    if (logoutVisible || sidebarVisible) break;
  }

  await page.waitForLoadState("networkidle");
  await page.waitForTimeout(1000);
}
